pilot/pkg/model/proxy_config_test.go

```diff
@@ -0,0 +1,360 @@
+					"c": "d",
+				},
+			},
+		},
+	}
+
+	for _, tc := range cases {
+		merged := mergeWithPrecedence(tc.first, tc.second)
+		if diff := cmp.Diff(merged, tc.expected); diff != "" {
+			t.Fatalf("expected and received not the same: %s", diff)
+		}
+	}
+}
+
+func TestEffectiveProxyConfig(t *testing.T) {
+	cases := []struct {
+		name          string
+		configs       []config.Config
+		defaultConfig *meshconfig.ProxyConfig
+		proxy         *NodeMetadata
+		expected      *meshconfig.ProxyConfig
+	}{
+		{
+			name: "CR applies to matching namespace",
+			configs: []config.Config{
+				newProxyConfig("ns", "test-ns",
+					&v1beta1.ProxyConfig{
+						Concurrency: v(3),
+					}),
+			},
+			proxy:    newMeta("test-ns", nil, nil),
+			expected: &meshconfig.ProxyConfig{Concurrency: v(3)},
+		},
+		{
+			name: "CR takes precedence over meshConfig.defaultConfig",
+			configs: []config.Config{
+				newProxyConfig("ns", istioRootNamespace,
+					&v1beta1.ProxyConfig{
+						Concurrency: v(3),
+					}),
+			},
+			defaultConfig: &meshconfig.ProxyConfig{Concurrency: v(2)},
+			proxy:         newMeta("bar", nil, nil),
+			expected:      &meshconfig.ProxyConfig{Concurrency: v(3)},
+		},
+		{
+			name: "workload matching CR takes precedence over namespace matching CR",
+			configs: []config.Config{
+				newProxyConfig("workload", "test-ns",
+					&v1beta1.ProxyConfig{
+						Selector: selector(map[string]string{
+							"test": "selector",
+						}),
+						Concurrency: v(3),
+					}),
+				newProxyConfig("ns", "test-ns",
+					&v1beta1.ProxyConfig{
+						Concurrency: v(2),
+					}),
+			},
+			proxy:    newMeta("test-ns", map[string]string{"test": "selector"}, nil),
+			expected: &meshconfig.ProxyConfig{Concurrency: v(3)},
+		},
+		{
+			name: "matching workload CR takes precedence over annotation",
+			configs: []config.Config{
+				newProxyConfig("workload", "test-ns",
+					&v1beta1.ProxyConfig{
+						Selector: selector(map[string]string{
+							"test": "selector",
+						}),
+						Concurrency: v(3),
+					}),
+			},
+			proxy: newMeta(
+				"test-ns",
+				map[string]string{
+					"test": "selector",
+				}, map[string]string{
+					annotation.ProxyConfig.Name: "{ \"concurrency\": 5 }",
+				}),
+			expected: &meshconfig.ProxyConfig{Concurrency: v(3)},
+		},
+		{
+			name: "CR in other namespaces get ignored",
+			configs: []config.Config{
+				newProxyConfig("ns", "wrong-ns",
+					&v1beta1.ProxyConfig{
+						Concurrency: v(1),
+					}),
+				newProxyConfig("workload", "wrong-ns",
+					&v1beta1.ProxyConfig{
+						Selector: selector(map[string]string{
+							"test": "selector",
+						}),
+						Concurrency: v(2),
+					}),
+				newProxyConfig("global", istioRootNamespace,
+					&v1beta1.ProxyConfig{
+						Concurrency: v(3),
+					}),
+			},
+			proxy:    newMeta("test-ns", map[string]string{"test": "selector"}, nil),
+			expected: &meshconfig.ProxyConfig{Concurrency: v(3)},
+		},
+		{
+			name: "multiple matching workload CRs all apply",
```

Comment:
Good point, I agree this should be consistent with other APIs. This will also mitigate some perf issues. Changed, and I will update the ProxyConfig API to describe that having multiple matching workload PCs is undefined.

Reply:
Changed namespace and mesh level merge semantics to do the same.